### PR TITLE
Improve how server/js client handle unexpected errors

### PR DESCRIPTION
**Pull request by freddyaboulton**
## Description

Noticed two things when I was debugging flaky tests:
- If there's an exception in the `queue/data/` route the client will hang forever. the server now catches general exceptions
- If the message didn't have an `event_id`, the client would error out, e.g. heartbeat, server_stopped. This wouldn't cause any noticeable errors to users (except console exceptions) but wanted to fix.My guess is that the playwright tests are mysteriously erroring server side so printing the exception message in the UI can help us debug.
- `gr.Info` and `gr.Warning` would also raise exceptions in the client. Would not be obvious to users but not desirable anyways.
- Think I found the issue for the weird HTTPExceptions in logs

## 🎯 PRs Should Target Issues

Before your create a PR, please check to see if there is [an existing issue](https://github.com/gradio-app/gradio/issues) for this change. If not, please create an issue before you create this PR, unless the fix is very small. 

Not adhering to this guideline will result in the PR being closed. 

## Tests

1. PRs will only be merged if tests pass on CI. To run the tests locally, please set up [your Gradio environment locally](https://github.com/gradio-app/gradio/blob/main/CONTRIBUTING.md) and run the tests: `bash scripts/run_all_tests.sh`

2. You may need to run the linters: `bash scripts/format_backend.sh` and `bash scripts/format_frontend.sh`

### CODE REVIEW

File: client/js/src/client.ts [hunk 1]
```diff
@@ -821,84 +822,129 @@ export function api_factory(
...
+										) {
+											if (event_callbacks[event_id]) {
+												delete event_callbacks[event_id];
+												if (Object.keys(event_callbacks).length === 0) {
+													close_stream();
+												}
 											}
 										}
+									} catch (e) {
+										console.error("Unexpected client exception", e);
```
Comment (freddyaboulton) [inline client/js/src/client.ts:930]: This should fix the `HTTPException` errors we keep seeing in the logs.

### Explanation
https://www.loom.com/share/5f7abfdb2f2143c58982f11dab2538b2?sid=d176fd1c-b5f5-43f5-be5f-f72af9b3abea

### Fix
https://www.loom.com/share/d6b9d87d05b64f69bfb8158ac565593f?sid=ca833ba1-4579-4536-8e14-50396ba3261c

**Comment by freddyaboulton** [comment]
Not sure how to test this actually. Should be good for review though!

**Comment by abidlabs** [comment]
Is this expected to resolve https://github.com/gradio-app/gradio/issues/6713?

**Comment by freddyaboulton** [comment]
@abidlabs Hard to know for sure since we don't know how to repro those issues. However, I did just verify that exceptions in user functions don't break the queue/app

https://www.loom.com/share/0b7d99654b444b6a8ff6d0c320c878ed?sid=bb08ee79-eb06-48f0-822d-c89d0819a1db

File: gradio/routes.py [hunk 2]
```diff
@@ -644,9 +650,17 @@ async def sse_stream(request: fastapi.Request):
                                     )
                                 ):
                                     return
-                except asyncio.CancelledError as e:
-                    del blocks._queue.pending_messages_per_session[session_hash]
-                    await blocks._queue.clean_events(session_hash=session_hash)
+                except Exception as e:
```
Comment (aliabid94) [inline gradio/routes.py:653]: Has to be BaseException to catch everything - asyncio.CancelledError is not an instance of Exception

File: client/js/src/client.ts [hunk 3]
```diff
@@ -821,84 +822,122 @@ export function api_factory(
...
+											endpoint: _endpoint,
+											fn_index,
+											time: new Date()
+										});
+										close_stream();
 									}
 								};
+								event_ids.push(event_id);
 								event_callbacks[event_id] = callback;
+								if (!stream_open) {
```
Comment (aliabid94) [inline client/js/src/client.ts:938]: why did you move this?

File: client/js/src/client.ts [hunk 4]
```diff
@@ -1014,6 +1053,12 @@ export function api_factory(
 				event_stream = new EventSource(url);
 				event_stream.onmessage = async function (event) {
 					let _data = JSON.parse(event.data);
+					if (_data.session_hash) {
+						await Promise.all(
+							event_ids.map((event_id) => event_callbacks[event_id](_data))
```
Comment (aliabid94) [inline client/js/src/client.ts:1058]: If this is the only reason we're keeping track of event_ids, we already have that list in the keys of event_callbacks. No need to create a second list right?

File: client/js/src/client.ts (hunk 3 above)
Comment (freddyaboulton) [reply client/js/src/client.ts:938]: Figured the stream should be opened after the callback got added

**Comment by aliabid94** [comment]
> If there's an exception in the queue/data/ route the client will hang forever. the server now catches general exceptions

What exceptions were happening?

File: client/js/src/client.ts (hunk 3 above)
Comment (aliabid94) [reply client/js/src/client.ts:938]: ok sure, but if a second request is sent while a first one is pending, the stream will be open from the first request, so we shouldn't be expecting it to be the case that the stream is receiving messages for an event only after post request is complete. 
Therefore, it can technically happen that the event stream starts sending messages for an event before the POST request to upload data finishes, and therefore the event callback to handle the message is not ready, which is a race condition that I do not know if we are hitting. I think I do resolve this in a PR I'm working on though

**Comment by freddyaboulton** [comment]
> What exceptions were happening?

We don't know because the client would hang forever. Although I think it's the reason the playwright tests were just hanging forever in the cases they failed.

File: client/js/src/client.ts (hunk 4 above)
Comment (aliabid94) [reply client/js/src/client.ts:1058]: also would prefer if the logic for targetting every event listener was more explicit than if the session_hash key is present. We don't use the session_hash value anyway. I think it'd make more sense that if an event_id is not provided, we trigger every event.

Comment (freddyaboulton) [reply client/js/src/client.ts]: Done!

Comment (freddyaboulton) [reply client/js/src/client.ts]: Also made the change about checking the absence of event_id to trigger all listeners!

**Review by aliabid94 — APPROVED** [review]
LGTM!

**Comment by freddyaboulton** [comment]
Thanks @aliabid94 !
